### ISSUE TITLE
Low accuracy for dataset in nsfw_data_scrapper

### ISSUE THREAD

**Issue by jiangyurong609** [GantMan/nsfw_model#22]
I tried to test the model against all the data from https://github.com/alexkimxyz/nsfw_data_scraper. It turns out the the model accuracy was pretty low (0.84) for porn classes. I checked several false negative images, should belong to porn. Any suggestions?

Below are detailed results for porn images:

porn 0.8378566785677277
sexy 0.09817904345614349
neutral 0.028440081768767722
hentai 0.033282149350465834
drawings 0.0022420468568952363

**Comment by GantMan** [comment]
I haven't had such poor results.   How clean/sure are you of your data?

Also which model did you use?   Keras or TF?   

I do not have such poor results on my machine, so I'm trying to gather the details.

**Comment by jiangyurong609** [comment]
I used this keras model https://s3.amazonaws.com/nsfwdetector/nsfw.299x299.h5. I basically ran the scripts in https://github.com/alexkimxyz/nsfw_data_scraper to generate the train dataset, and used the train dataset to test it. There are 106153 porn images tested with the model. What's your suggestions around this test? Thanks

**Comment by GantMan** [comment]
Wow.  That result sucks!  It sounds like I need to re-run the scraper and get some more data and see if I can retrain the model.   It's doing amazing with my local dataset.  Which was pulled not long ago.

Feel free to see if training on your dataset improves performance.

**Comment by jiangyurong609** [comment]
@GantMan thanks, I will try to do it when I have time. Meanwhile, if you had chance to re-train, please let me know. Thanks

**Comment by GantMan** [comment]
Also, if you run the "self clense" script can you let me know if you have a lot of errors in your dataset?

Mine was pretty clean.  From what I recall, a fresh pull on NSFW data scraper can be pretty off.

**Comment by TechnikEmpire** [comment]
I've been training and experimenting with neural networks like this and the problem is there will always be "low accuracy" because of whole context classification. You cannot train a whole context neural network for pornography classification that isn't poor, its not possible.

You need a SSD to do this properly.

**Comment by jiangyurong609** [comment]
@TechnikEmpire what's the advantage of SSD for this simple classification?

**Comment by TechnikEmpire** [comment]
"Simple classification" tells me that you don't understand the complexity of the problem so my answer won't be understood either.

**Comment by GantMan** [comment]
Hrmmm.  That came off kinda harsh. This project is about everything and everyone.  Would you like to try again?   I think this is a great chance for you to practice sharing your research in a friendly yet challenging way.

**Comment by TechnikEmpire** [comment]
Yep I come off as harsh.

I've been developing various models and methods for this task since 2014. There is always shortcomings because the human mind has endless imagination when generating adult content. I have a squeezenet model on github that beats Yahoo's open source model on pornographic images, but both my model and Yahoo's fail miserably on many other forms of pornography.

If you feed either model full frontal, 90 degree rotated images they fail (despite doing this preprocessing during training).

They both fail horribly at detecting oral (think most clothes are on).

All of these failures are rooted in whole/single context classification. An ssd trained to identify the objects that unequivocally represent sexuality alone (breasts, genitalia) is the only possible way to generate a truly universally highly accurate model.

I'm in the process of rewriting darknet in c++AMP for this purpose. My goal is to distribute such a model in broadly available desktop software.

There are ways to rapidly generate ssd training data. 3D scans of these things. Then write 3D software that renders these scans at every possible angle while generating regions of interest. Then further generate more variants for training with translation and composition.

Its not a simple problem was my point, that's all.

**Comment by TechnikEmpire** [comment]
"Write 3d software" sounds extensive. To clarify, I wrote an add-on for blender. I may write a simple D3D program in the future for greater control. Anyway, I hope my remarks are helpful.
